hail/src/main/scala/is/hail/expr/ir/AggOp.scala

```diff
@@ -4,30 +4,72 @@ import is.hail.expr.ir.agg.Extract
 import is.hail.expr.types.physical.PType
 import is.hail.expr.types.virtual._
 
+object AggStateSignature {
+  def apply(sig: AggSignature): AggStateSignature = AggStateSignature(Map(sig.op -> sig), sig.op)
+}
+
+case class AggStateSignature(m: Map[AggOp, AggSignature], default: AggOp, nested: Option[Seq[AggStateSignature]] = None) {
+  lazy val defaultSignature: AggSignature = m(default)
+  lazy val resultType: Type = Extract.getResultType(this)
+  def lookup(op: AggOp): AggSignature = m(op)
+
+  def toCanonicalPhysical: AggStatePhysicalSignature = AggStatePhysicalSignature(
+    m.map { case (op, sig) =>
+      (op, sig.toPhysical(sig.initOpArgs.map(PType.canonical), sig.seqOpArgs.map(PType.canonical)))
+    },
+    default,
+    nested.map(_.map(_.toCanonicalPhysical))
+  )
+}
+
+
 case class AggSignature(
   op: AggOp,
   initOpArgs: Seq[Type],
-  seqOpArgs: Seq[Type],
-  nested: Option[Seq[AggSignature]] = None) {
-  lazy val returnType: Type = agg.Extract.getResultType(this)
+  seqOpArgs: Seq[Type]) {
+
   def toPhysical(initOpTypes: Seq[PType], seqOpTypes: Seq[PType]): PhysicalAggSignature = {
-    assert(nested.isEmpty)
-    (initOpTypes, initOpArgs).zipped.foreach { case (pt, t) => assert(pt.virtualType == t) }
-    (seqOpTypes, seqOpArgs).zipped.foreach { case (pt, t) => assert(pt.virtualType == t) }
-    PhysicalAggSignature(op, initOpTypes, seqOpTypes, None)
+    (initOpTypes, initOpArgs).zipped.foreach { case (pt, t) => assert(pt.virtualType isOfType t) }
+    (seqOpTypes, seqOpArgs).zipped.foreach { case (pt, t) => assert(pt.virtualType isOfType t) }
+    PhysicalAggSignature(op, initOpTypes, seqOpTypes)
   }
+
+  lazy val singletonContainer: AggStateSignature = AggStateSignature(Map(op -> this), op, None)
+
+  // only to be used with virtual non-nested signatures on ApplyAggOp and ApplyScanOp
+  lazy val returnType: Type = AggStateSignature(this).resultType
```

Comment:
should this have an assert (match AggStateSignature(this).nested { case Some => throw } )

Reply:
The AggStateSignature(AggSignature) constructor has nowhere to get any nested aggregators for, so I think this is probably OK. Perhaps I should assert that the op isn't Group() or AggElements() or AggElementsLengthCheck(), but we'll get a good match error when we try to do this anyway, I think.